### PR TITLE
Friendship v2

### PR DESCRIPTION
**Pull request by geraldgsh**
- Created 2 way friend request feature
- Created 2 way friend confirmation feature
- Created 2 way friend deletion & friend request cancellation
- Created Rspec test cases for the above

### CODE REVIEW

File: app/models/friendship.rb [hunk 1]
```diff
@@ -3,4 +3,26 @@
 class Friendship < ApplicationRecord
```
[the remark this answers is not in the document]
Comment (adriaanbd) [reply app/models/friendship.rb]: In addition to the above, you could also explore the `validates_uniqueness_of` as a way of validating the relationship and preventing a user to friend itself. Check the documentation for it [here](https://apidock.com/rails/ActiveRecord/Validations/ClassMethods/validates_uniqueness_of). or [hre](https://api.rubyonrails.org/classes/ActiveRecord/Validations/ClassMethods.html#method-i-validates_uniqueness_of). Look into how you could use `scope` or `conditions`. Does this give you any further ideas? 

[`OPTIONAL`]

[the remark this answers is not in the document]
Comment (adriaanbd) [reply app/models/friendship.rb]: Your schema file could also index `user_id` and `friend_id` together with a `unique: true`, here:
https://github.com/geraldgsh/societalbook/blob/3af1fcb4d10d9b599fbe2097efe0486641dc43aa/db/schema.rb#L26

Something like this:

`t.index ["user_id", "friend_id"], name: "index_friendships_on_user_id_and_friend_id", unique: true`

[`OPTIONAL`]

[the remark this answers is not in the document]
Comment (adriaanbd) [reply app/models/friendship.rb]: You could also add your own validation, let's say `validate :uniqueness_of_friendship` and then create a method that returns a boolean:

```
def uniqueness_of_friendship
    if friendship_exists?
        self.errors.add(:unique_friendship, 'Already friends!')
    elsif trying_to_friend_oneself?
        self.errors.add(:invalid_friendship, ''Can't friend yourself")
end
```

[`OPTIONAL`]

**Comment by geraldgsh** [comment]
# Missing Pack folder in public folder

Pack folder that exist on our local machine did not find it way in Github repo despite being pushed.  Hence @adriaanbd was not able to run app.

image

So the folder was uploaded to Repo manually by drag and drop;

image

File: app/models/friendship.rb [hunk 2]
```diff
@@ -3,4 +3,26 @@
...
   belongs_to :user
   belongs_to :friend, :class_name => "User"
-end
+
+  def self.create_friendship(user_id, friend_id)
+    user_friendship = Friendship.create(user_id: user_id, friend_id: friend_id, status: false)
+    friend_friendship = Friendship.create(user_id: friend_id, friend_id: user_id, status: false)
```
[the remark this answers is not in the document]
Comment (geraldgsh) [reply app/models/friendship.rb]: 1. Added "disallow_self_friendship" method in friendship model to validate against friend ownself
2. Added "duplicate_check" method in friendship model to validate against duplicate friend requests

**Comment by geraldgsh** [comment]
@adriaanbd Many thanks 🙏 for your feedback as well as suggestions. It helped us to refine our code.

**Comment by geraldgsh** [comment]
Thanks @Redvanisation for the review.

We are unable to use composite primary keys because it breaks the SQL transaction.

**Comment by adriaanbd** [comment]
Congratulations team, good work on pushing this through!

**Pull request merged**